[aotd] Do not force contiguous() for channels_last

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135225

Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata

Subclasses logic:
 - Previously coercing tangents logic did not handle nested subclasses case, fixing this.

For Subclasses we deduce memory format for subclass_tensor first, then for each element of subclass:
[subclass_tensor_memory_format, subclass_tensor_elem0_memory_format, ... ]

If subclass element (__tensor_flatten__[0] tensors) is also subclass => on its place we will have a nested list of the same structure.

The recursive traversal of subclass tree is expensive. So we do memory format deduction and coercing at the same time, to keep only one traverse for this. With this approach there  is no regression in comparison with previous logic which also does one traversal. (`coerce_tangent_and_suggest_memory_format` method). 



Other small change:
Remove duplicated not-related comment.


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```

Benchmarking:
After change:
```
└─ $ PYTORCH_AOTD_DEBUG_PROFILE=1 python test/functorch/test_aotdispatch.py -k test_benchmark_grads_no_force_contiguous
Benchmark SUBCLASS avg_bwd_duration:4.059906005859375 ms
Benchmark NO_SUBCLASS avg_bwd_duration:3.1563830375671387 ms
```
Before change:
```
BEFORE_CHANGE SUBCLASS 4.1194
```

No siginificant changes in processing time.

(We do single traverse of subclass tree for collecting memory_formats and coercing during tracing.)

IvanKobzarev committed:
[aotd] Do not force contiguous() for channels_last

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"




[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -6022,6 +6022,101 @@ def fn(x):
+            s.backward()
+
+        contiguous_events = [
+            event for event in prof.events() if event.name == "aten::contiguous"
```
hmm yeah, I can't think of a better way to test than what you're doing (running the profiler and making sure there is not actually a contiguous() call at runtime).

Just for sanity, did you confirm that this test fails without your change?

bdhirsh replied:
also, a comment explaining the purpose of this test would be very appreciated

IvanKobzarev replied:
Yes, It fails without changes in AOTD.

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -6022,6 +6022,101 @@ def fn(x):
+                    memory_format=torch.channels_last
+                ),
+            )
+            [inp.retain_grad() for inp in ret]
```
why are all of the retain_grad calls necessary throughout these tests?

IvanKobzarev replied:
Yeah, I am not doing grad assert checks here, we can remove it.

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

Chillee commented on test/functorch/test_aotdispatch.py:
```diff
@@ -6022,6 +6022,108 @@ def fn(x):
         out = torch.compile(fn, backend="aot_eager", fullgraph=True)(inp)
         self.assertEqual(ref_out, out)
 
+    def test_channels_last_grads_no_force_contiguous_dense(self):
```
I would add some more tests with some outputs that are channels last, and some outputs that aren't.

Chillee commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -63,10 +63,14 @@
 # on all tangents at runtime.
 # In the future, you could imagine lifting this restriction, since these contiguous()
 # calls can have noticeable perf overhead depending on the model.
```
update this comment

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -665,11 +669,27 @@ def view_avoid_dupes_with_primals(t):
+            suggest_memory_format = torch._prims_common.suggest_memory_format
+            if is_traceable_wrapper_subclass(t):
+                return [
+                    recursive_suggest_memory_format(getattr(t, attr))
```
I'm a bit torn because:

(1) right now you are trying to track the memory formats of inner tensors separately from their outer subclasses. This technically isn't enough, because you need to handle nested subclasses (what if `getattr(t, attr)` is another tensor subclass that you need to recursively flatten and get the memory formats of its inner tensors?)

(2) you could fix that by having some recursive nested lists that contain the full hierarchy of suggested memory formats for every subclass tangent. I'm a bit worried about traversing those nested lists being bad for hot-path in the backward.

Alternatively, we could make the simplifying assumption that if a subclass advertises as "channels-last contiguous", it will likely also have channels-last-contiguous inner tensors (and vis versa), and therefore only bother to do this bookkeeping on the top-level subclass and not on the inner tensors.

bdhirsh replied:
I guess it is totally possible to construct a subclass that violates that, though, and the recursive tracking shouldn't be *that* slow. So what do you think of:

(1) do the recursive tracking

(2) run a quick microbenchmark that times the backward of a function like this, and make sure its backward isn't a lot slower
```
def create_inp(x):
    return TwoTensor(TwoTensor(x.clone(), x.clone()), TwoTensor(x.clone(), x.clone()))

@torch.compile
def f(*args):
    return args

x = torch.randn(1, requires_grad=True)
inps = [create_inp(x) for _ in range(100)]
outs = f(inps)
# you probably just want to measure the overhead of the `CompiledFunction.backward()` directly, so you don't measure overhead from the rest of autograd
sum(outs).sum().backward()

```

IvanKobzarev replied:
I think (maybe not in the first version) I did a recursive handling for subclasses attributes.
Will add test for NestedSubclasses to verify it.

IvanKobzarev replied:
Benchmarked:
```
        benchmark_inps = [inps_fn(torch.randn(2, 3, 5, 5, requires_grad=True).to(memory_format=torch.channels_last)) for _ in range(100)]

        bwd_total_duration = 0
        for inps in benchmark_inps:
            outs = torch.compile(mc, backend="aot_eager", fullgraph=True)(*inps)
            s = outs[0].sum()
            time_start = time.time()
            s.backward()
            bwd_duration = time.time() - time_start
            bwd_total_duration += bwd_duration
        
        avg_bwd_duration = bwd_total_duration / len(benchmark_inps)
        print(f"XXX    SUBCLASS_GROUP avg_bwd_duration:{avg_bwd_duration*1000} ms")
        
        class M2(torch.nn.Module):
            def __init__(self) -> None:
                super().__init__()
                self.conv = torch.nn.Conv2d(3, 3, 3)

            def forward(self, x0, x1, x2, x3):
                return self.conv(x0), self.conv(x1), self.conv(x2), self.conv(x3)

        m2 = M2()
        m2.to(memory_format=torch.channels_last)
        m2.train()

        def inps_fn2(x):
            return (
                x.clone(),
                x.clone(),
                x.clone(),
                x.clone()
            )

        benchmark_inps2 = [inps_fn2(torch.randn(2, 3, 5, 5, requires_grad=True).to(memory_format=torch.channels_last)) for _ in range(100)]

        bwd_total_duration = 0
        for inps in benchmark_inps2:
            outs = torch.compile(m2, backend="aot_eager", fullgraph=True)(*inps)
            s = outs[0].sum()
            time_start = time.time()
            s.backward()
            bwd_duration = time.time() - time_start
            bwd_total_duration += bwd_duration
        
        avg_bwd_duration = bwd_total_duration / len(benchmark_inps)
        print(f"XXX NO_SUBCLASS_GROUP avg_bwd_duration:{avg_bwd_duration * 1000} ms")
```

XXX    SUBCLASS_GROUP avg_bwd_duration:7.817392349243163 ms
XXX NO_SUBCLASS_GROUP avg_bwd_duration:3.940465450286865 ms
Subclasses bwd is 2 times slower avg

bdhirsh replied:
oh that delta between subclass and no delta is a good datapoint (that's also probably telling us that other things like the subclass `flatten/unflatten` we do in the subclass path adds up to quite a bit of overhead in general?)

My specific question though was more about if this PR causes any regressions for the existing subclass path. So "SUBCLASS_GROUP" (with your changes) vs. "SUBCLASS_GROUP" (without your changes)

IvanKobzarev replied:
Measured:

Without changes:
channels_last: 11.92ms

With changes:
contiguous: 10.68ms
channels_last: 8ms

No regression with changes, used all-contiguous() to remove the difference of having contiguous() without-changes. But still do not understand fully why it speeds up the backward in case of all contiguous.

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -6022,6 +6022,162 @@ def fn(x):
+                        torch.randn(2, 3, 5, 5, requires_grad=True).to(
+                            memory_format=torch.channels_last
+                        ),
+                        torch.randn(2, 3, 5, 5, requires_grad=True).to(
```
even better, let's make the inner two tensors have a mix of channels-first-contiguous and channels-last-contiguous, and ensure that we **still** don't have any runtime contiguous calls (so we "guessed all of our memory formats" properly)

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -55,6 +55,16 @@
 
+# Workaround of https://github.com/pytorch/pytorch/issues/62027
+# tensor.contiguous() guarantees to return non-zero sorted-stride,
+# tensor.to(memory_format=torch.contiguous_format) can keep zero strides.
```
hmm... can you talk more about why you think it's worth trying to handle the zero strides case? 

The common case where I imagine zero strides showing up is when you compile a model (not including the loss), and then run `.sum().backward()`. The autograd engin will do something like `torch.ones(1).expand(tangent_shape)`, and the incoming tangent input to the backward will have a zero stride.

We are still kind of out-of-luck in this case though, because the strides of the tangent will be **different** than the strides of the forward graph output (the tangent technically has overlapping memory, while the forward output does not).

We will have been forced to trace out a backward graph ahead-of-time that assumed that the tangent was a plain contiguous tensor, so it does kind of feel like we are forced to emit the contiguous call in the backward.

Is the zero-stride-handling here attempting to handle that case? Or something else

IvanKobzarev replied:
Copy from chat:

Inductor generates asserts for sizes and strides, and as output strides were non-zero - then it generates non-zero assert.
When at runtime tangents come with zero stride and .to(contiguous) does not change them - it fails on inductor assert.

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -314,6 +314,8 @@ class ViewAndMutationMeta:
     # This list is generated after calling make_runtime_safe().
     traced_tangent_metas: Optional[List[Any]] = None
 
+    traced_tangent_memory_formats: Optional[List[Any]] = None
```
A comment on the structure here would be nice. In particular, I think the new type after our discussion is something like this?
```
Optional[List[Union[MemoryFormat, List[MemoryFormat]]]

# for each tangent at index i:
#     if the tangent is a plain tensor, traced_tangent_memory_formats[i] holds the memory format of the tangent that we need to coerce to
#     if the tangent is a subclass, traced_tangent_memory_formats[i] holds a list of memory formats, containing the expected memory format of the subclass **and** all of its inner tensors
```

what order is that flattened list stored in? in-order tree traversal?

IvanKobzarev replied:
Yes, will update it.

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/subclass_utils.py:
```diff
@@ -347,3 +347,11 @@ def compute_inner_mutated_inp_indices_from_subclass_meta(
+
+
+def subclass_setattr(s, attr, value):
+    # setattr on Subclass could fail with AttributeError e.g. NestedTensor for _min_seqlen_tensor
```
cc @jbschlosser does this ring a bell (or sound suspicious)?

IvanKobzarev replied:
Updated PR with removing it, avoiding to do setattr() as _min_seqlen_tensor is already contiguous.

But technically yes, it's a good design question for subclasses, if framework should do some actions on each subclass subelement indepedently or only use Subclass Authors API or __torch_dispatch__ calls on it only.

As we can shift Subclass to some inconsistent state that Authors did not expect :)

IvanKobzarev replied:
In dynamo execution `setattr(NestedTensor(), "_min_seqlen_tenor", value)` fails.

bdhirsh replied:
from offline - we realized that NJT returning `_min_seqlen_tensor` as an "inner tensor attribute" but having it secretly be a computed property that reads from a cache is a bit suspicious

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```


[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -91,16 +99,62 @@ def coerce_tangent(x):
+        for attr in attrs:
             elem = getattr(out, attr)
-            if not elem.is_contiguous():
-                elem_contig = elem.contiguous()
```
fyi @tugsbayasgalan - looks like our contiguous coercion logic wasn't technically running on nested subclasses (pretty rare case that I doubt people have hit yet) - @IvanKobzarev is technically bundling that fix into this PR.

side note @IvanKobzarev - can you update the PR description so it's clear what you're actually fixing here (since there are two independing bugfixes in this PR)

IvanKobzarev commented on torch/_functorch/_aot_autograd/input_output_analysis.py:
```diff
@@ -82,18 +87,6 @@ def remove_dupe_metadata(
     )
 
 
-# Given our ViewAndMutation metadata, this fn constructs a new set of metadata,
```
Yes, it was duplicated and not related to this PR :)

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -669,13 +723,20 @@ def view_avoid_dupes_with_primals(t):
+                if (output_tangents_start_idx <= i and i < output_tangents_end_idx)
+                else torch.contiguous_format,
+            )
+            for i, tt in enumerate(traced_tangents)
```
hmm, we used to always run `coerce_tangents` on all `traced_tangents`. Why are you filtering out some elements now? (if I'm reading this properly, I think it is trying to apply this logic to `f_input_tangents` and `f_output_tangents` but not `intermediate_bases`)

IvanKobzarev replied:
Here the logic is that for 
`f_output_tangents` we deduce memory format with suggest_memory_format and coerce to that deduced memory_format.
`f_input_tangents`, `intermediate_bases` - the logic is as previous, we skip suggest memory format, forcing that format be contiguous

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -91,16 +99,62 @@ def coerce_tangent(x):
+    if not out.is_contiguous(memory_format=memory_format):
+        was = out
+        out = out.contiguous(memory_format=memory_format)
+        updated = out is not was
```
nit: if you are checking `out is not was` here, then you don't need the `is_contiguous()` check above: you can just unconditionally call `contiguous()`, and see if it returned the input directly

IvanKobzarev replied:
Yeah, was not sure about guarantees of matching is_contiguous/contiguous. But if they match and .contiguous() guaranteed to  return unchanged if is_contiguous() - then we can do it.

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -91,16 +99,62 @@ def coerce_tangent(x):
+):
+    updated = False
+    if not isinstance(x, Tensor):
+        return x, None, updated
```
this fn feels a bit complicated (it returns `[Tensor, MemoryFormat, bool]`?) - can you add return types to the fn and document them?

bdhirsh replied:
reading this more - I think returning the `was_updated` bool is kind of confusing, since it is ignored in most places. What if instead, we:

- don't return a `was_updated` bool
- in the places where the caller needed it, they can just do the check themselves to know if contiguous had an effect:
```
updated_out, mem_format = coerce_tangent_and_suggest_memory_format(out)
if updated_out is not out:
    setattr(...)
```

bdhirsh commented on torch/_functorch/_aot_autograd/input_output_analysis.py:
```diff
@@ -49,12 +49,16 @@ def remove_dupe_metadata(
         if keep_arg_mask[m.mutated_inp_runtime_indices[i]]
     ]
     traced_tangents = filtered_inp_traced_tangents + other_traced_tangents
+    assert m.traced_tangent_memory_formats is not None
```
side note (since I see the logic you're changing here is in the `remove_dupe_metadata` codepath): we are pretty sure that this is dead code in torch.compile, since dynamo has its own logic to remove dupe'd inputs.

@jamesjwu tried to kill this code a while back (but I think ran into some issues?) - we probably want to give it another shot at some point https://github.com/pytorch/pytorch/pull/127306

bdhirsh commented on torch/_functorch/_aot_autograd/input_output_analysis.py:
```diff
@@ -49,12 +49,16 @@ def remove_dupe_metadata(
+    assert m.traced_tangent_memory_formats is not None
+    traced_tangent_memory_formats = [torch.contiguous_format] * len(
+        filtered_inp_traced_tangents
+    ) + m.traced_tangent_memory_formats[num_data_mutations:]
```
meh yeah i guess this is good enough (not 100% accurate, but again this should be dead code)

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -56,19 +56,27 @@
-def coerce_tangent(x):
+# If runtime specfied tangents will not have the same memory format as predicted traced tangents,
+# we coerce them at runtime to traced tangents memory format.
+def coerce_tangent(x, memory_format=torch.contiguous_format):
```
hmm i don't think I understand - why do we have both `coerce_tangent()` and `coerce_tangent_and_suggest_memory_format()` floating around? It seems like your new util should subsume the old one (and we can replace call sites of the first with the second?)

IvanKobzarev replied:
I introduced a new one to not regress the coerce_tangent that is also called from input_and_mutation_aliases.
coerce_tangent_and_suggest_memory_format is doing more things calling suggest_memory_format and optional logic of force_memory_format.
So I decided to keep the fast version in parallel. (It's callsed from input_output_analysis)

bdhirsh replied:
Hmm, the only place I see it used in `input_output_analysis.py` is in `create_synthetic_base_metadata`, which is called at compile time anyway. So I don't think this fn is actually used anywhere at runtime?

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -91,16 +99,62 @@ def coerce_tangent(x):
+# Coercing and collecting traced tangents memory format in one recursive traversal
+# mypy: ignore-errors
+def coerce_tangent_and_suggest_memory_format(
+    x: Tensor, force_memory_format: Optional[torch.memory_format] = None
```
hmm... do we actually pass in a non-None value for `force_memory_format` anywhere? I don't think I see one.

Given that we have a separate coercion function for trace-time vs runtime, the "force_memory_format" seems unnecessary? (the trace time fn never forces, the runtime fn always forces)

IvanKobzarev replied:
Yes, non-None value means that we need to run suggest_memory_formt to deduce it, this happens for all output tangents. While for others - aliases, mutation input tangents - it will have a value.

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1868,24 +1878,46 @@ def get_types_for_tangents(tangents):
-                ]
+                    all_args = [
+                        (
+                            AOTDispatchAutograd.coerce_runtime_tangent_tracing_memory_format(
```
hmm why do we need to call `coerce_runtime_tangent_tracing_memory_format` separately on both sides of the branch above?

IvanKobzarev replied:
For the subclasses case we want to call it before flattening. Potentially we can move coercing before subclasses branch I will try.

IvanKobzarev commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -314,6 +314,15 @@ class ViewAndMutationMeta:
+    #   if the tangent is a subclass, traced_tangent_memory_formats[i] holds a list of memory formats,
+    #     containing the expected memory format of the subclass **and** all of its inner tensors
+    traced_tangent_memory_formats: Optional[
+        List[Union[torch.memory_format, List[torch.memory_format]]]
```
Yes, this typing is wrong, we have here recursive typing 
```
TMF = Union[torch.memory_format, List[TMF]]
traced_tangent_memory_formats: Optional[List[TMF]]
```
will try to express it with typing, or replace with Any and put the recursive typing in comment

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata

Subclasses logic:
 - Previously coercing tangents logic did not handle nested subclasses case, fixing this.

For Subclasses we deduce memory format for subclass_tensor first, then for each element of subclass:
[subclass_tensor_memory_format, subclass_tensor_elem0_memory_format, ... ]

If subclass element (__tensor_flatten__[0] tensors) is also subclass => on its place we will have a nested list of the same structure.

The recursive traversal of subclass tree is expensive. So we do memory format deduction and coercing at the same time, to keep only one traverse for this. With this approach there  is no regression in comparison with previous logic which also does one traversal. (`coerce_tangent_and_suggest_memory_format` method). 



Other small change:
Remove duplicated not-related comment.


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```

Benchmarking:
After change:
```
└─ $ PYTORCH_AOTD_DEBUG_PROFILE=1 python test/functorch/test_aotdispatch.py -k test_benchmark_grads_no_force_contiguous
Benchmark SUBCLASS avg_bwd_duration:4.059906005859375 ms
Benchmark NO_SUBCLASS avg_bwd_duration:3.1563830375671387 ms
```
Before change:
```
BEFORE_CHANGE SUBCLASS 4.1194
```

No siginificant changes in processing time.

(We do single traverse of subclass tree for collecting memory_formats and coercing during tracing.)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata

Subclasses logic:
 - Previously coercing tangents logic did not handle nested subclasses case, fixing this.

For Subclasses we deduce memory format for subclass_tensor first, then for each element of subclass:
[subclass_tensor_memory_format, subclass_tensor_elem0_memory_format, ... ]

If subclass element (__tensor_flatten__[0] tensors) is also subclass => on its place we will have a nested list of the same structure.

The recursive traversal of subclass tree is expensive. So we do memory format deduction and coercing at the same time, to keep only one traverse for this. With this approach there  is no regression in comparison with previous logic which also does one traversal. (`coerce_tangent_and_suggest_memory_format` method). 



Other small change:
Remove duplicated not-related comment.


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```

Benchmarking:
After change:
```
└─ $ PYTORCH_AOTD_DEBUG_PROFILE=1 python test/functorch/test_aotdispatch.py -k test_benchmark_grads_no_force_contiguous
Benchmark SUBCLASS avg_bwd_duration:4.059906005859375 ms
Benchmark NO_SUBCLASS avg_bwd_duration:3.1563830375671387 ms
```
Before change:
```
BEFORE_CHANGE SUBCLASS 4.1194
```

No siginificant changes in processing time.

(We do single traverse of subclass tree for collecting memory_formats and coercing during tracing.)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Do not force contiguous() for channels_last"


Original Issue: https://github.com/pytorch/pytorch/issues/134644

We assume trace_tangents to have the same memory_format as inputs, outputs, intermediate during first tracing.

=>
Tracing time:
- Store trace_tangents_memory_formats in metadata
- Coerce tangents to deduced memory_format

Runtime:
- Coerce tangents to tracing memory format from metadata

Subclasses logic:
 - Previously coercing tangents logic did not handle nested subclasses case, fixing this.

For Subclasses we deduce memory format for subclass_tensor first, then for each element of subclass:
[subclass_tensor_memory_format, subclass_tensor_elem0_memory_format, ... ]

If subclass element (__tensor_flatten__[0] tensors) is also subclass => on its place we will have a nested list of the same structure.

The recursive traversal of subclass tree is expensive. So we do memory format deduction and coercing at the same time, to keep only one traverse for this. With this approach there  is no regression in comparison with previous logic which also does one traversal. (`coerce_tangent_and_suggest_memory_format` method). 



Other small change:
Remove duplicated not-related comment.


Testing

```
python test/functorch/test_aotdispatch.py -k test_channels_last_grads_no_force_contiguous
```

Benchmarking:
After change:
```
└─ $ PYTORCH_AOTD_DEBUG_PROFILE=1 python test/functorch/test_aotdispatch.py -k test_benchmark_grads_no_force_contiguous
Benchmark SUBCLASS avg_bwd_duration:4.059906005859375 ms
Benchmark NO_SUBCLASS avg_bwd_duration:3.1563830375671387 ms
```
Before change:
```
BEFORE_CHANGE SUBCLASS 4.1194
```

No siginificant changes in processing time.

(We do single traverse of subclass tree for collecting memory_formats and coercing during tracing.)

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev commented:
@pytorchbot merge
